### PR TITLE
Echo Client Example

### PR DESCRIPTION
**Pull request by Akvinikym**
An example of how to write an Echo protocol client in C++ is added with this PR. 

Also, some compatibility issues were fixed; they have appeared when testing the client against a Go server.

### CODE REVIEW

File: example/01-echo/libp2p_echo_client.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  auto injector =
+      libp2p::injector::makeHostInjector(libp2p::injector::useKeyPair(keypair));
+  auto host = injector.create<std::shared_ptr<libp2p::Host>>();
+
+  // create Echo protocol object - it implement the logic of both server and
+  // client, but in this example it's used as a client-only
+  libp2p::protocol::Echo echo{libp2p::protocol::EchoConfig{1}};
+
+  // create Host; we use it to create Echo stream
+  auto context = injector.create<std::shared_ptr<boost::asio::io_context>>();
```
Comment (kamilsa) [inline example/01-echo/libp2p_echo_client.cpp]: can we please logic of creating host into separate function?

File: example/01-echo/libp2p_echo_client.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  using libp2p::common::operator""_unhex;
+
+  // this keypair generates a PeerId
+  // "12D3KooWLs7RC93EGXZzn9YdKyZYYx3f9UjTLYNX1reThpCkFb83"
+  KeyPair keypair{PublicKey{{Key::Type::Ed25519,
+                             "a4249ea6d62bdd8bccf62257ac4899ff284796"
+                             "3228b388fda288db5d64e517e0"_unhex}},
+                  PrivateKey{{Key::Type::Ed25519,
+                              "4a9361c525840f7086b893d584ebbe475b4ec"
+                              "7069951d2e897e8bceb0a3f35ce"_unhex}}};
```
Comment (igor-egorov) [inline example/01-echo/libp2p_echo_client.cpp]: We do not need this keypair to run as a client.

File: example/01-echo/libp2p_echo_client.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+  // create a default Host via an injector, overriding a random-generated
+  // keypair with ours
+  auto injector =
+      libp2p::injector::makeHostInjector(libp2p::injector::useKeyPair(keypair));
```
Comment (igor-egorov) [inline example/01-echo/libp2p_echo_client.cpp]:
```suggestion
      libp2p::injector::makeHostInjector();
```

File: example/01-echo/libp2p_echo_client.cpp (hunk 1 above)
Comment (igor-egorov) [reply example/01-echo/libp2p_echo_client.cpp]: the verb is missing)

File: example/01-echo/libp2p_echo_client.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  // client, but in this example it's used as a client-only
+  libp2p::protocol::Echo echo{libp2p::protocol::EchoConfig{1}};
+
+  // create Host; we use it to create Echo stream
+  auto context = injector.create<std::shared_ptr<boost::asio::io_context>>();
+  context->post([host{std::move(host)}, &echo] {
+    // this address is to be substituted with an address of server
+    auto server_ma_res = libp2p::multi::Multiaddress::create(
+        "/ip4/127.0.0.1/tcp/40010/ipfs/"
+        "12D3KooWLs7RC93EGXZzn9YdKyZYYx3f9UjTLYNX1reThpCkFb83");
```
Comment (igor-egorov) [inline example/01-echo/libp2p_echo_client.cpp]: Can we get rid of hardcoded value?
Probably we could read the whole address string from cin.

File: src/muxer/yamux/yamuxed_connection.cpp [hunk 5]
```diff
@@ -40,7 +40,7 @@ namespace libp2p::connection {
         connection_{std::move(connection)},
         config_{config} {
     // client uses odd numbers, server - even
-    last_created_stream_id_ = connection_->isInitiator() ? 1 : 0;
+    last_created_stream_id_ = connection_->isInitiator() ? 1 : 2;
```
Comment (igor-egorov) [inline src/muxer/yamux/yamuxed_connection.cpp]: why?

Comment (Akvinikym) [reply src/muxer/yamux/yamuxed_connection.cpp]: StreamIds start from 1, not from 0, so the previous variant was incorrect

**Pull request merged**